objectstore-types/src/lib.rs

```diff
@@ -203,3 +203,36 @@ impl Metadata {
         Ok(headers)
     }
 }
+
+/// Newtype around [`String`] that can protect against certain cases of
+/// logging sensitive values. Please use this instead of `String` in
+/// things like configuration structs which may be dumped blindly into
+/// debug logs.
+#[derive(Deserialize, Clone, Serialize)]
```

Comment:
As an idea, what if we also implement `Display` like so, to ensure it can never be accidentally added?

```
impl fmt::Display for InternalId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if cfg!(debug_assertions) {
            unimplemented!("do not display secrets");
        } else {
            write!(f, "[redacted]")
        }
    }
}
```

Reply:
You surely mean "implement it as a panic" so that noone will implement it any other way?
Because otherwise, by not implementing it at all, the compiler and type system will make sure noone will accidentally add a *usage* of the trait.

Reply:
> so that noone will implement it any other way

Exactly, that's the idea. It does come with the unfortunate downside that now Display is in fact implemented. Maybe it's enough to trust future reviewers, though.